compiler/rustc_passes/src/liveness.rs

```diff
@@ -515,6 +515,33 @@ struct Liveness<'a, 'tcx> {
     cont_ln: HirIdMap<LiveNode>,
 }
 
+// As we go through the code, we warn about unreachable code after expressions
+// with uninhabited types. We do not want to warn if the last expression of an
+// if or match arm has an uninhabited type, though, because the code after the
+// if/match could still be reachable via another arm -- unless all arms
```

Comment:
Hm, can you say more about the specific case being addressed here? It's not quite clear to me why diverges would be yes and warn unreachable would be no, yet we shouldn't lint on that expression?

Reply:
```rust
enum Void {}
fn void() -> Void {todo!()}

fn bar(b: bool) {
    if b {
        void(); // A
    } else {
        // B
    }
    // C
}
```
In this example, `A` diverges because it has an uninhabited type. We can't issue a warning that "any code [in particular, `C`] following this expression is unreachable", though, because `C` might be reachable via the `else` branch. Hence `A` diverges and we don't want to warn about unreachable successors (`C`). If `B` also turns out to be diverging, we will warn about the entire `if` expression diverging and making `C` unreachable.

Reply:
Hm, so it seems like when we're exiting the if in the CFG, we should only consider C diverging if A *and* B have diverged -- so at point C we should have diverges = no, and as such not emit any warnings?

Reply:
Exactly.

Reply:
My comment was meant to suggest that I (still) don't understand the purpose/necessity for WarnUnreachable -- it seems to me that we should already not be warning when seeing C just based on Diverges = no at that point, so WarnUnreachable isn't actually necessary?

Reply:
Ah, sorry. Consider another example then:
```rust
if ... {
    diverges(); // X
    alsoDiverges(); // Y
}
// Z
```
Y has the successor Z and diverges, so Diverges=Yes, but WarnUnreachable=No because Z is also reachable via the (implicit) else branch. By contrast, X has the successor Y and diverges, so Diverges=Yes, and WarnUnreachable=Yes because X is Y's only predecessor. If there was a non-implicit else branch, the Diverges information would be used to check whether the entire `if` is diverging (in which case we would want to warn about Z being unreachable).

So you are right in that WarnUnreachable doesn't make a difference if Diverges=No, but it does for Diverges=Yes.

Reply:
I would expect the implicit else branch to have Diverges::No as its "result", making point Z have the same diverges as point A (presumably No, in your example):

```
// A
if ... {
   diverges(); // X
   diverges(); // Y
}
// Z
```

Y's successor is Z, but Z has two predecessors in the CFG: (roughly) A or Y, depending on if the if was taken. (Really, the predecessor of Z is the if condition and Y, I guess).

Getting this right is likely to be relatively hard, which is why I'm not sure we should be trying to roll this on our own -- a MIR rewrite would make it much easier.

Reply:
> Getting this right is likely to be relatively hard, which is why I'm not sure we should be trying to roll this on our own -- a MIR rewrite would make it much easier.

I totally agree, but I don't have the MIR knowledge and time to do this myself at the moment.